ports/esp32s2/common-hal/busio/I2C.c

```diff
@@ -0,0 +1,225 @@
+
+    common_hal_mcu_delay_us(10);
+
+    gpio_pulldown_dis(sda->number);
+    gpio_pulldown_dis(scl->number);
+
+    // We must pull up within 3us to achieve 400khz.
+    common_hal_mcu_delay_us(3);
+
+    if (gpio_get_level(sda->number) == 0 || gpio_get_level(scl->number) == 0) {
+        reset_pin_number(sda->number);
+        reset_pin_number(scl->number);
+        mp_raise_RuntimeError(translate("SDA or SCL needs a pull up"));
+    }
+#endif
+
+
+    if (xSemaphoreCreateBinaryStatic(&self->semaphore) != &self->semaphore) {
+        mp_raise_RuntimeError(translate("Unable to create lock"));
+    }
+    xSemaphoreGive(&self->semaphore);
+    self->sda_pin = sda;
+    self->scl_pin = scl;
+    self->i2c_num = I2C_NUM_MAX;
+    for (i2c_port_t num = 0; num < I2C_NUM_MAX; num++) {
+        if (i2c_status[num] == STATUS_FREE) {
+            self->i2c_num = num;
+        }
+    }
+    if (self->i2c_num == I2C_NUM_MAX) {
+        mp_raise_ValueError(translate("All I2C peripherals are in use"));
+    }
+    i2c_status[self->i2c_num] = STATUS_IN_USE;
+    i2c_config_t i2c_conf = {
+        .mode = I2C_MODE_MASTER,
+        .sda_io_num = self->sda_pin->number,
+        .scl_io_num = self->scl_pin->number,
+        .sda_pullup_en = GPIO_PULLUP_DISABLE,  /*!< Internal GPIO pull mode for I2C sda signal*/
+        .scl_pullup_en = GPIO_PULLUP_DISABLE,  /*!< Internal GPIO pull mode for I2C scl signal*/
+
+        .master = {
+            .clk_speed = frequency,
+        }
+    };
+    esp_err_t result = i2c_param_config(self->i2c_num, &i2c_conf);
+    if (result != ESP_OK) {
+        mp_raise_ValueError(translate("Invalid pins"));
+    }
+    result = i2c_driver_install(self->i2c_num,
+                                I2C_MODE_MASTER,
+                                0,
+                                0,
+                                0);
+    if (result != ESP_OK) {
+        mp_raise_OSError(MP_EIO);
+    }
+
+    claim_pin(sda);
+    claim_pin(scl);
+}
+
+bool common_hal_busio_i2c_deinited(busio_i2c_obj_t *self) {
+    return self->sda_pin == NULL;
+}
+
+void common_hal_busio_i2c_deinit(busio_i2c_obj_t *self) {
+    if (common_hal_busio_i2c_deinited(self)) {
+        return;
+    }
+
+    i2c_driver_delete(self->i2c_num);
+    i2c_status[self->i2c_num] = STATUS_FREE;
+
+    common_hal_reset_pin(self->sda_pin);
+    common_hal_reset_pin(self->scl_pin);
+    self->sda_pin = NULL;
+    self->scl_pin = NULL;
+}
+
+bool common_hal_busio_i2c_probe(busio_i2c_obj_t *self, uint8_t addr) {
+    i2c_cmd_handle_t cmd = i2c_cmd_link_create();
+    i2c_master_start(cmd);
+    i2c_master_write_byte(cmd, addr << 1, true);
+    i2c_master_stop(cmd);
+    esp_err_t result = i2c_master_cmd_begin(self->i2c_num, cmd, 10);
+    i2c_cmd_link_delete(cmd);
+    return result == ESP_OK;
+}
+
+bool common_hal_busio_i2c_try_lock(busio_i2c_obj_t *self) {
+    self->has_lock = xSemaphoreTake(&self->semaphore, 0) == pdTRUE;
+    return self->has_lock;
+}
```

Comment:
If already self->has_lock, it gets lost.

Here's what samd does:
```
bool common_hal_busio_i2c_try_lock(busio_i2c_obj_t *self) {
    bool grabbed_lock = false;
    CRITICAL_SECTION_ENTER()
        if (!self->has_lock) {
            grabbed_lock = true;
            self->has_lock = true;
        }
    CRITICAL_SECTION_LEAVE();
    return grabbed_lock;
}
```

As for internal object pointers, a problem will only arise if the object is locked _WHEN_ it becomes long-lived.  I'm comfortable that this could be a "don't do that, then".

Reply:
Ok, I now return false if this object already has the lock.